sw/airborne/modules/guidance/gvf_parametric/gvf_parametric.cpp

```diff
@@ -0,0 +1,455 @@
+}
+
+static void send_circle_parametric(struct transport_tx *trans, struct link_device *dev)
+{
+  uint32_t now = get_sys_time_msec();
+  uint32_t delta_T = now - gvf_parametric_t0;
+
+  if(delta_T < 200)
+    if (gvf_parametric_trajectory.type == ELLIPSE_3D) {
+      pprz_msg_send_CIRCLE(trans, dev, AC_ID, &gvf_parametric_trajectory.p_parametric[0], &gvf_parametric_trajectory.p_parametric[1], &gvf_parametric_trajectory.p_parametric[2]);
+  }
+}
+
+#endif // PERIODIC TELEMETRY
+
+#ifdef __cplusplus
+}
+#endif
+
+void gvf_parametric_init(void)
+{
+  gvf_parametric_control.w = 0;
+  gvf_parametric_control.delta_T = 0;
+  gvf_parametric_control.s = 1;
+  gvf_parametric_control.k_roll = GVF_PARAMETRIC_CONTROL_KROLL;
+  gvf_parametric_control.k_climb = GVF_PARAMETRIC_CONTROL_KCLIMB;
+  gvf_parametric_control.k_psi = GVF_PARAMETRIC_CONTROL_KPSI;
+  gvf_parametric_control.L = GVF_PARAMETRIC_CONTROL_L;
+  gvf_parametric_control.beta = GVF_PARAMETRIC_CONTROL_BETA;
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_GVF_PARAMETRIC, send_gvf_parametric);
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_CIRCLE, send_circle_parametric);
+#endif
+
+}
+
+void gvf_parametric_set_direction(int8_t s)
+{
+  gvf_parametric_control.s = s;
+}
+
+void gvf_parametric_control_2D(float kx, float ky, float f1, float f2, float f1d, float f2d, float f1dd, float f2dd)
+{
+
+  uint32_t now = get_sys_time_msec();
+  gvf_parametric_control.delta_T = now - gvf_parametric_t0;
+  gvf_parametric_t0 = now;
+
+  if (gvf_parametric_control.delta_T > 300) { // We need at least two iterations for Delta_T
+    gvf_parametric_control.w = 0; // Reset w since we assume the algorithm starts
+    return;
+  }
+
+  float L = gvf_parametric_control.L;
+  float beta = gvf_parametric_control.beta*gvf_parametric_control.s;
+
+  Eigen::Vector3f X;
+  Eigen::Matrix3f J;
+
+  // Error signals phi_x and phi_y
+  struct EnuCoor_f *pos_enu = stateGetPositionEnu_f();
+  float x = pos_enu->x;
+  float y = pos_enu->y;
+
+  float phi1 = L * (x - f1);
+  float phi2 = L * (y - f2);
+
+  gvf_parametric_trajectory.phi_errors[0] = phi1; // Error signals for the telemetry
+  gvf_parametric_trajectory.phi_errors[1] = phi2;
+
+  // Chi
+  X(0) = L*beta*f1d - kx*phi1;
+  X(1) = L*beta*f2d - ky*phi2;
+  X(2) = L + beta*(kx*phi1*f1d + ky*phi2*f2d);
+  X *= L;
+
+  // Jacobian
+  J.setZero();
+  J(0, 0) = -kx*L;
+  J(1, 1) = -ky*L;
+  J(2, 0) = (beta*L)*(beta*f1dd+kx*f1d);
+  J(2, 1) = (beta*L)*(beta*f2dd+ky*f2d);
+  J(2, 2) = beta*beta*(kx*(phi1*f1dd-L*f1d*f1d) + ky*(phi2*f2dd-L*f2d*f2d));
+  J *= L;
+
+  // Guidance algorithm
+  float ground_speed = stateGetHorizontalSpeedNorm_f();
+  float w_dot = (ground_speed * X(2)) / sqrtf(X(0) * X(0) + X(1) * X(1));
+
+  Eigen::Vector3f xi_dot;
+  struct EnuCoor_f *vel_enu = stateGetSpeedEnu_f();
+  float course = stateGetHorizontalSpeedDir_f();
+
+  xi_dot << vel_enu->x, vel_enu->y, w_dot;
```

Comment:
Interesting. For sure it works on board because we have flown already with the code. I do not know what Eigen is doing exactly here. Maybe they have overloaded the operator << for the initialization and it is doing just a[0] = 1, a[1] = 2, etc. I do not use the operator << for anything else, so maybe this is why it does not trigger any problems.

All matrices and vectors have a fixed/harcoded predefined size, and we compile with "EIGEN_NO_MALLOC". So, concerning Eigen, nothing is dynamic here.

Reply:
ok, then maybe it was a configuration error on my side